amanuensis/resources/admin/admin_associated_user.py

```diff
@@ -17,18 +18,97 @@
 
 
 def update_role(project_id, user_id, email, role):
-    with flask.current_app.db.session as session:  
+    with flask.current_app.db.session as session:
         ret = udm.update_associated_users(session, project_id, user_id, email, role)
         return ret
 
+
 def add_associated_users(users):
+
+    users = [defaultdict(lambda: "", user) for user in users]
+
     with flask.current_app.db.session as session:
         associated_user_schema = AssociatedUserSchema(many=True)
         ret = []
+        users_with_id, users_with_email = [], []
         for user in users:
-            ret.append(udm.add_associated_user(session, user["project_id"], user["email"] if "email" in user else None, user["id"] if "id" in user else None))
-        associated_user_schema.dump(ret)
-        return ret
+            if "email" in user:
+                users_with_email.append(user)
+            elif "id" in user:
+                users_with_id.append(user)
+            else:
+                raise UserError(
+                    "The associated user must have at least an email or id. Neither was found."
+                )
+        user_emails = [user["email"] for user in users_with_email]
+        user_ids = [user["id"] for user in users_with_id]
 
+        associated_users = udm.get_associated_users(session, user_emails)
+        associated_users += udm.get_associated_users_by_id(session, user_ids)
 
+        fence_users = []
+        if user_emails:
+            fence_users.extend(
+                fence.fence_get_users(config, usernames=user_emails)["users"]
+            )
+        if user_ids:
+            fence_users.extend(fence.fence_get_users(config, ids=user_ids)["users"])
+        if fence_users:
+            fence_users = [
+                {"email": user["name"], "id": user["id"]} for user in fence_users
+            ]
 
+        user_missing = []  # users not found in associated users table
+
+        associated_user_emails = [a_user.email for a_user in associated_users]
+
+        for user in users_with_email:
+            if user["email"] not in associated_user_emails:
+                user_missing.append(user)
+
+        associated_user_ids = [a_user.user_id for a_user in associated_users]
+
+        for user in users_with_id:
+            if user["id"] not in associated_user_ids:
+                user_missing.append(user)
```

Comment:
comment for myself to check - this assumes that the current records have both id, and email otherwise it can create duplicates. We only have 4 users now, so check it